src/controllers/companyController.js

```diff
@@ -0,0 +1,87 @@
+import { companies } from "../models/index.js";
+import notFoundError from "../erros/notFoundError.js";
+
+class CompanyController {
+  static async registerCompany(req, res, next) {
+    try {
+      const newCompany = await companies.create(req.body);
+      res.status(201).json({ message: "Loja criada!", company: newCompany });
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async listCompanyById(req, res, next) {
+    try {
+      const id = req.params.id;
+      const foundCompany = await companies.findById(id);
+
+      if (foundCompany !== null) {
+        res.status(200).json(foundCompany);
+      } else {
+        next(new notFoundError("Id do livro não localizado"));
+      }
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async updateCompany(req, res, next) {
+    try {
+      const id = req.params.id;
+      const foundCompany = await companies.findById(id);
+
+      if (foundCompany !== null) {
+        const id = req.params.id;
+        const updatedCompany = await companies.findByIdAndUpdate(id, req.body, {
+          runValidators: true,
+          new: true,
+        });
+        res
+          .status(200)
+          .json({ message: "Loja atualizada!", company: updatedCompany });
+      } else {
+        next(new notFoundError("Id do livro não localizado"));
+      }
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async listCompanies(req, res, next) {
+    try {
+      const search = await buildSearchQuery(req.query);
+
+      if (search !== null) {
+        const searchResults = companies.find(search);
+
+        req.result = searchResults;
+
+        next();
+      } else {
+        res.status(200).send([]);
+      }
+    } catch (err) {
+      next(err);
+    }
+  }
+
+  static async countCompanies() {
+    const companiesFound = await companies.find();
+    return companiesFound.length;
+  }
+}
+
+async function buildSearchQuery(params) {
+  const { name, city, document } = params;
+  
+  let search = {};
+
+  if (name) search.name = { $regex: name, $options: "i" };
+  if (city) search.city = city;
+  if (document) search.document = { $regex: document, $options: "i" };
+
+  return search;
+}
```

Comment:
Fabinho, tenho alguns pontos sobre esse método:


- Esse método não precisa ser async, pois ele não lida com nenhum processo assíncrono dentro dele, apenas faz o tratamento com dados de um objeto. Logo, você pode remover o `async` dele, e não precisa do `await` onde ele for chamado.

- Você começou muito bem o mapeamento dos campos da query, mas um dos requisitos da consulta é mapear também os parâmetros de paginação `from` e `limit`, que podem (ou não) ser mapeados pelo cliente.

Uma dica que te dou é: quando for montar essa searchQuery,  você pode distinguir o que são parâmetros de paginação e o que são parâmetros de filtros. Logo, teu método pode ficar assim:

```suggestion
function buildSearchQuery(params) {
  const { from, limit, name, city, document } = params;

  const search = {
    pagination: {
      from: 0,
      limit: 20,
    },
    filter: {},
  };

  if (from) search.pagination.from = Number(from);
  if (limit) search.pagination.limit = Number(limit);
  if (name) search.filter.name = { $regex: name, $options: "i" };
  if (city) search.filter.city = city;
  if (document) search.filter.document = { $regex: document, $options: "i" };

  return search;
}
```

Ai você me pergunta: mas por que os parâmetros de paginação tem valores default? E a resposta é simples: tornar a API um pouco mais segura para lidar com dados. Imagina um contexto onde tu tenha um milhão de registro de companies, e tu faça uma consulta sem informar nenhum parâmetro de paginação. Já imaginou o caos que é para o cliente receber uma resposta de um milhão de dados de uma vez? Isso se a API não quebrar pelo alto volume de dados retornados.

Para mitigar isso, você pode definir valores default de paginação. Assim você assegura que, caso o usuário não passe nenhum parâmetro de paginação, o retorno padrão da API é seguro. Logo:

- o `from` com o default 0 significa que, caso o usuário não o informe na query, a API irá retornar os documentos a partir do primeiro documento encontrado em diante.
- o `limit` com default 20 significa que, caso o usuário não o informe na query, a API irá retornar apenas os primeiros 20 documentos encontrados.

Já o filter, conforme eu te disse antes, ele pode ser vazio sem problemas. Isso significa que, se eu não informar nenhum parâmetro de filtragem, eu quero que a API retorne os 'x' primeiros documentos que ela tiver, independente dos parâmetros que possuam.

Fazendo dessa forma, lá no método `listCompanies`, a consulta pode ser feita da seguinte forma:

```javascript
      const search = buildSearchQuery(req.query);

      const data = await companies
        .find(search.filter) // os filtros da query
        .limit(search.pagination.limit) // o limite da paginação
        .skip(search.pagination.from); // no mongoose o 'skip' tem o mesmo significado do 'from'

      const total = await companies
        .countDocuments(search.filter); // na contagem de documentos não consideramos paginação
```

Reply:
> Você começou muito bem o mapeamento dos campos da query, mas um dos requisitos da consulta é mapear também os parâmetros de paginação from e limit, que podem (ou não) ser mapeados pelo cliente.

Os parâmetros de paginação podem (inclusive com valores-padrão) serem inseridos na query pelo usuário, quem cuida disso é o middleware paginate.js 🙂

Reply:
> Esse método não precisa ser async, pois ele não lida com nenhum processo assíncrono dentro dele, apenas faz o tratamento com dados de um objeto. Logo, você pode remover o async dele, e não precisa do await onde ele for chamado.

Bem notado!